ember-client/shared/src/main/scala/org/http4s/ember/client/EmberClientBuilder.scala

```diff
@@ -380,6 +382,156 @@ final class EmberClientBuilder[F[_]: Async: Network] private (
+      }
+    }
+  
+  def buildWebSocket(webSocketKey: Key[Socket[F]]): Resource[F, Client[F]] =
+    for {
+      sg <- Resource.pure(sgOpt.getOrElse(Network[F]))
+      tlsContextOptWithDefault <-
+        tlsContextOpt
+          .fold(Network[F].tlsContext.systemResource.attempt.map(_.toOption))(
+            _.some.pure[Resource[F, *]]
+          )
+      builder =
+        KeyPool.Builder
+          .apply[F, RequestKey, EmberConnection[F]]((requestKey: RequestKey) =>
+            EmberConnection(
+              org.http4s.ember.client.internal.ClientHelpers
+                .requestKeyToSocketWithKey[F](
+                  requestKey,
+                  tlsContextOptWithDefault,
+                  checkEndpointIdentification,
+                  serverNameIndication,
+                  sg,
+                  additionalSocketOptions,
+                ),
+              chunkSize,
+            ) <* Resource
+              .eval(logger.trace(s"Created Connection - RequestKey: ${requestKey}"))
+              .onFinalize(
+                logger.trace(
+                  s"Shutting Down Connection - RequestKey: ${requestKey}"
+                )
+              )
+          )
+          .withDefaultReuseState(Reusable.DontReuse)
+          .withIdleTimeAllowedInPool(idleTimeInPool)
+          .withMaxPerKey(maxPerKey)
+          .withMaxTotal(maxTotal)
+          .withOnReaperException(_ => Applicative[F].unit)
+      pool <- builder.build
+      optH2 <- (if (enableHttp2) tlsContextOptWithDefault else None).traverse { context =>
+        H2Client.impl[F](
+          pushPromiseSupport.getOrElse { case (_, _) => Applicative[F].pure(Outcome.canceled) },
+          context,
+          unixSockets,
+          logger,
+          if (pushPromiseSupport.isDefined) default
+          else
+            default.copy(enablePush = H2Frame.Settings.SettingsEnablePush(false)),
+          checkEndpointIdentification,
+          serverNameIndication,
+        )
+      }
+    } yield {
+      def webClient(request: Request[F]): Resource[F, Response[F]] =
+        for {
+          managed <- ClientHelpers.getValidManaged(pool, request)
+          _ <- Resource.eval(
+            pool.state.flatMap { poolState =>
+              logger.trace(
+                s"Connection Taken - Key: ${managed.value.keySocket.requestKey} - Reused: ${managed.isReused} - PoolState: $poolState"
+              )
+            }
+          )
+          responseResource <- Resource.makeCaseFull((poll: Poll[F]) =>
+            poll(
+              ClientHelpers
+                .request[F](
+                  request,
+                  managed.value,
+                  chunkSize,
+                  maxResponseHeaderSize,
+                  idleConnectionTime,
+                  timeout,
+                  userAgent,
+                )
+            )
+          ) { case ((response, drain), exitCase) =>
+            exitCase match {
+              case Resource.ExitCase.Succeeded =>
```

Comment:
Likewise, if we have a pattern-match as the body of a case, we can flatten the pattern-matches by nesting the cases: 

```suggestion
          ) { 
            case ((response, drain), Resource.ExitCase.Succeeded) =>
```

Reply:
Thanks for all the suggestions!